kernel/src/net/socket/ip/raw/mod.rs

```diff
@@ -0,0 +1,427 @@
+    },
+    match_sock_option_mut,
+    net::{
+        iface::IfaceEx,
+        socket::{
+            options::{Error as SocketError, IpHdrIncl, SocketOption},
+            util::{
+                options::{IpSocketOptionSet, SocketOptionSet},
+                send_recv_flags::SendRecvFlags,
+                socket_addr::SocketAddr,
+                MessageHeader,
+            },
+            Socket,
+        },
+    },
+    prelude::*,
+    process::signal::{PollHandle, Pollable, Pollee},
+    util::{net::Protocol, MultiRead, MultiWrite},
+};
+
+mod bound;
+mod unbound;
+
+#[derive(Debug, Clone)]
+struct OptionSet {
+    socket: SocketOptionSet,
+    ip: IpSocketOptionSet,
+}
+
+impl OptionSet {
+    fn new() -> Self {
+        let socket = SocketOptionSet::new_raw();
+        let ip = IpSocketOptionSet::new_raw();
+        OptionSet { socket, ip }
+    }
+}
+
+pub struct RawDGramSocket {
+    options: RwLock<OptionSet>,
+    inner: RwLock<Takeable<Inner>, LocalIrqDisabled>,
+    nonblocking: AtomicBool,
+    pollee: Pollee,
+}
+
+enum Inner {
+    Unbound(UnBoundRaw),
+    Bound(BoundRaw),
+}
+
+impl Inner {
+    fn bind(
+        self,
+        addr: &IpAddress,
+        can_reuse: bool,
+    ) -> core::result::Result<BoundRaw, (Error, Self)> {
+        let unbound_raw = match self {
+            Inner::Unbound(unbound_raw) => unbound_raw,
+            Inner::Bound(bound_raw) => {
+                return Err((
+                    Error::with_message(Errno::EINVAL, "the socket is already bound to an address"),
+                    Inner::Bound(bound_raw),
+                ));
+            }
+        };
+
+        let bound_raw = match unbound_raw.bind(addr, can_reuse) {
+            Ok(bound_raw) => bound_raw,
+            Err((err, unbound_raw)) => return Err((err, Inner::Unbound(unbound_raw))),
+        };
+        Ok(bound_raw)
+    }
+
+    fn bind_to_ephemeral_endpoint(
+        self,
+        remote_endpoint: &IpEndpoint,
+    ) -> core::result::Result<BoundRaw, (Error, Self)> {
+        if let Inner::Bound(bound_raw) = self {
+            return Ok(bound_raw);
+        }
+
+        let endpoint = get_ephemeral_endpoint(remote_endpoint);
+        self.bind(&endpoint.addr, false)
+    }
+}
+
+impl RawDGramSocket {
+    pub fn new(nonblocking: bool, protocol: Protocol) -> Arc<Self> {
+        Arc::new_cyclic(|me| {
+            let unbound_raw = UnBoundRaw::new(me.clone() as _, protocol);
+            Self {
+                inner: RwLock::new(Takeable::new(Inner::Unbound(unbound_raw))),
+                nonblocking: AtomicBool::new(nonblocking),
+                pollee: Pollee::new(),
+                options: RwLock::new(OptionSet::new()),
+            }
+        })
+    }
+
+    pub fn is_nonblocking(&self) -> bool {
+        self.nonblocking.load(Ordering::SeqCst)
+    }
+
+    pub fn set_nonblocking(&self, nonblocking: bool) {
+        self.nonblocking.store(nonblocking, Ordering::SeqCst);
```

Comment:
`is_nonblocking` should be accessed with `Ordering::Relaxed` unless you want to synchronize something.